### PR TITLE
INT: Auto-import unresolved types in generated code

### PR DESCRIPTION
**Pull request by mchernyavsky**
Auto-import unresolved types in intentions:
  - Implement members
  - Specify type explicitly
  - Fill match arms
  - Add remaining patterns
  - Use destructuring declaration

This PR doesn't affect:
  - Add missing fields / Recursively add missing fields
  - Initialize with a default value

because `RsDefaultValueBuilder.buildFor` can't use unresolved types.

Based on https://github.com/intellij-rust/intellij-rust/pull/3099.
Closes https://github.com/intellij-rust/intellij-rust/issues/3586. Closes https://github.com/intellij-rust/intellij-rust/issues/3495.
Relates to https://github.com/intellij-rust/intellij-rust/pull/3520.

### CODE REVIEW

**Comment by Undin** [comment]
@mchernyavsky Could you rebase it onto master, please?

**Comment by mchernyavsky** [comment]
@Undin Done.

**Comment by Undin** [comment]
bors try

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 1]
```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
...
-    is RsPatStruct,
-    is RsPatTupleStruct -> PathParsingMode.COLONS
-    else -> PathParsingMode.NO_COLONS
+/**
+ * @param toImport  Set of unresolved items that can be imported
+ * @param toQualify Set of unresolved items that can't be imported
+ */
+data class TypeReferencesInfo(
+    val toImport: Set<RsQualifiedNamedElement>,
+    val toQualify: Set<RsQualifiedNamedElement>
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: Why do we need this list?
Looks like we don't use it anywhere

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 2]
```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
...
+                is TyAdt ->
+                    result.add(ty.item)
+                is TyAnon ->
+                    result.addAll(ty.traits.map { it.element })
+                is TyTraitObject ->
+                    result.add(ty.trait.element)
+                is TyProjection -> {
+                    result.add(ty.trait.element)
+                    result.add(ty.target)
+                }
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: I think it can be reformatted into 
```kotlin
is TyAdt -> result.add(ty.item)
is TyAnon -> result.addAll(ty.traits.map { it.element })
is TyTraitObject -> result.add(ty.trait.element)
is TyProjection -> {
    result.add(ty.trait.element)
    result.add(ty.target)
}
```

Comment (Undin) [reply src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: And maybe it's worth to use `+=` instead of `add` as more kotlin way.
But it's a matter of taste so I don't insist

File: src/main/kotlin/org/rust/ide/intentions/DestructureIntention.kt [hunk 3]
```diff
@@ -47,6 +49,9 @@ class DestructureIntention : RsElementBaseIntentionAction<DestructureIntention.C
     }
 
     private fun handleStruct(project: Project, editor: Editor, patIdent: RsPatIdent, struct: RsStructItem) {
+        val (toImport, _) = patIdent.getTypeReferencesInfoFromTy(struct.declaredType)
+        patIdent.importElements(toImport)
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/intentions/DestructureIntention.kt]: Hm, looks like we always do the same actions: call `getTypeReferencesInfoFrom*` and call `importElements`
Maybe we can create a single public function to make both actions inside it?

Comment (Undin) [reply src/main/kotlin/org/rust/ide/intentions/DestructureIntention.kt]: Also, I don't like that current functions are extention functions of `RsElement` because we use element only as context.
So I'd prefer pass element as explicit argument.
If you don't want to create new top-level function(s), let's create some object class like `RsImportHelper` and move them into it

Comment (Undin) [reply src/main/kotlin/org/rust/ide/intentions/DestructureIntention.kt]: And maybe it's worth to call `import` method after main action?
Imagine the following case: our new awesome import feature throws an exception. And in this case, main action won't be executed and user doesn't see any change. Otherwise, we make all necessary actions and just miss imports that user can import manually in case of potential exception

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 4]
```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
...
 }
 
-private val RsPath.pathParsingMode: PathParsingMode get() = when (parent) {
-    is RsPathExpr,
-    is RsStructLiteral,
-    is RsPatStruct,
-    is RsPatTupleStruct -> PathParsingMode.COLONS
-    else -> PathParsingMode.NO_COLONS
+/**
+ * @param toImport  Set of unresolved items that can be imported
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]:
```suggestion
 * @param toImport  Set of unresolved items that should be imported
```

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 5]
```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
...
+ * - Items that can't be imported
+ */
+private fun processRawImportSubjects(
+    context: RsElement,
+    rawImportSubjects: Set<RsQualifiedNamedElement>
+): TypeReferencesInfo {
+    val importSubjects = hashMapOf<String, MutableSet<RsQualifiedNamedElement>>()
+    for (element in rawImportSubjects) {
+        val name = element.name ?: continue
+        val group = importSubjects.getOrPut(name) { hashSetOf() }
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]:
```suggestion
        importSubjects.getOrPut(name, ::hashSetOf) += element
```

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 6]
```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
...
+ * Takes `rawImportSubjects` and filters items that are unresolved in the current `context`.
+ * Then splits this items into two sets:
+ * - Items that can be imported
+ * - Items that can't be imported
+ */
+private fun processRawImportSubjects(
+    context: RsElement,
+    rawImportSubjects: Set<RsQualifiedNamedElement>
+): TypeReferencesInfo {
+    val importSubjects = hashMapOf<String, MutableSet<RsQualifiedNamedElement>>()
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: Why do we need this map at all?
Can't we just process all elements from `rawImportSubjects` directly?

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 7]
```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
...
+    for (element in rawImportSubjects) {
+        val name = element.name ?: continue
+        val group = importSubjects.getOrPut(name) { hashSetOf() }
+        group.add(element)
+    }
+
+    val toQualifiedName = hashSetOf<RsQualifiedNamedElement>()
+    processNestedScopesUpwards(context, TYPES) { entry ->
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: If `importSubjects` is not necessary, `org.rust.lang.core.psi.ext.RsItemElementKt#filterInScope` maybe useful here

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 8]
```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
...
+private fun collectImportSubjectsFromTy(
+    ty: Ty,
+    subst: Substitution,
+    result: MutableSet<RsQualifiedNamedElement>
+) {
+    ty.substitute(subst).visitWith(object : TypeVisitor {
+        override fun visitTy(ty: Ty): Boolean {
+            when (ty) {
+                is TyAdt ->
+                    result.add(ty.item)
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: This code doesn't take into account type aliases. It should do it at least for `SpecifyTypeExplicitlyIntention` because I've just merged #4403
For example, 
```rust
use crate::foo::bar_baz;

mod foo {
    pub struct Bar<T>(T);
    pub struct Baz;
    pub type BarBaz = Bar<Baz>;

    pub fn bar_baz() -> BarBaz { Bar(Baz) }
}

fn main() {
    let bar/*caret*/ = bar_baz();
}
```

**Comment by mchernyavsky** [comment]
@Undin Fixed.

File: src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt [hunk 9]
```diff
@@ -142,6 +142,12 @@ class RsPsiFactory(
         return structLiteralField
     }
 
+    fun createStructLiteralFieldFull(name: String, value: String): RsStructLiteralField {
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/psi/RsPsiFactory.kt]: It seems it's unused

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 10]
```diff
@@ -671,20 +675,170 @@ data class ImportContext private constructor(
...
     }
 }
 
-private val RsPath.pathParsingMode: PathParsingMode get() = when (parent) {
-    is RsPathExpr,
-    is RsStructLiteral,
-    is RsPatStruct,
-    is RsPatTupleStruct -> PathParsingMode.COLONS
-    else -> PathParsingMode.NO_COLONS
+object RsImportHelper {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: Maybe move this class to separate file?

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 11]
```diff
@@ -576,6 +526,60 @@ fun ImportCandidate.import(context: RsElement) {
     insertionScope.insertUseItem(psiFactory, "$prefix${info.usePath}")
 }
 
+// Semantic signature of method is `ImportItem.canBeImported(mod: RsMod)`
+// but in our case `mod` is always same and `mod` needs only to get set of its super mods
+// so we pass `superMods` instead of `mod` for optimization
+fun QualifiedNamedItem.canBeImported(superMods: LinkedHashSet<RsMod>): ImportInfo? {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: Why did you move it and make public?

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 12]
```diff
@@ -671,20 +675,170 @@ data class ImportContext private constructor(
...
+        }
+
+        return TypeReferencesInfo(importSubjects.flatMap { it.value }.toSet(), toQualifiedName)
+    }
+
+    /**
+     * @param toImport  Set of unresolved items that should be imported
+     * @param toQualify Set of unresolved items that can't be imported
+     */
+    data class TypeReferencesInfo(
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: Can be private

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 13]
```diff
@@ -671,20 +675,170 @@ data class ImportContext private constructor(
...
+     */
+    private fun processRawImportSubjects(
+        context: RsElement,
+        rawImportSubjects: Set<RsQualifiedNamedElement>
+    ): TypeReferencesInfo {
+        val importSubjects = hashMapOf<String, MutableSet<RsQualifiedNamedElement>>()
+        for (element in rawImportSubjects) {
+            val name = element.name ?: continue
+            importSubjects.getOrPut(name, ::hashSetOf) += element
+        }
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]: Could you come up with case when value set has size > 1?

Comment (mchernyavsky) [reply src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt]:
```
use crate::a::foo;

mod a {
    mod b { pub struct S; }
    mod c { pub struct S; }
    pub fn foo() -> (b::S, c::S) { unimplemented!() }
}

fn main() {
    let x/*caret*/ = foo();
}
```

**Comment by mchernyavsky** [comment]
@Undin Fixed.
